src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,310 @@
+    }
+
+    @Override
+    public ItemResource withPatch(DataComponentPatch patch) {
+        if (isEmpty() || patch.isEmpty() || innerStack.getComponentsPatch().equals(patch))
+            return this;
+
+        ItemStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+
+        return ItemResource.of(stack);
+    }
+
+    @Override
+    public <D> ItemResource with(DataComponentType<D> type, D data) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.set(type, data);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
+
+        return ItemResource.of(stack);
+    }
+
+    //This is overridden to return ItemResource to allow method chaining
+    @Override
+    public <D> ItemResource with(Supplier<? extends DataComponentType<D>> type, D data) {
+        return with(type.get(), data);
+    }
+
+    @Override
+    public ItemResource without(DataComponentType<?> type) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.remove(type);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
+
+        return ItemResource.of(stack);
+    }
+
+    //This is overridden to return ItemResource to allow method chaining
+    @Override
+    public ItemResource without(Supplier<? extends DataComponentType<?>> type) {
+        return without(type.get());
+    }
+
+    @Override
+    public DataComponentMap getComponents() {
+        return innerStack.immutableComponents();
+    }
+
+    @Override
+    public DataComponentPatch getComponentsPatch() {
+        return innerStack.getComponentsPatch();
+    }
+
+    /**
+     * Creates an {@link ItemStack} of the specified count.
+     *
+     * @param count The amount of the item the stack should have. Must be non-negative.
+     * @return A new copy of the inner item stack with the specified count.
+     * @throws IllegalArgumentException when count is negative.
+     */
+    public ItemStack toStack(int count) {
+        TransferPreconditions.checkNonNegative(count);
+        if (count == 0) return ItemStack.EMPTY;
+        return this.innerStack.copyWithCount(count);
+    }
+
+    /**
+     * Creates an {@link ItemStack} with a count of 1.
+     *
+     * @return A new copy of the inner item stack with a count of 1.
+     */
+    public ItemStack toStack() {
+        return this.innerStack.copyWithCount(1);
+    }
+
+    public int getMaxStackSize() {
+        return innerStack.getMaxStackSize();
+    }
+
+    /**
+     * Returns the hover name of the {@link ItemStack}.
+     *
+     * @return The hover name of the {@link ItemStack}
+     */
+    public Component getHoverName() {
+        return innerStack.getHoverName();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        return obj instanceof ItemResource other && other.is(innerStack);
```

Comment:
Isn't it standard practice with final classes to just do a class equality check instead of an `instanceof`?